agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,700 @@
+	if !a.isPrimary {
+		pcaType = acmpca.CertificateAuthorityTypeSubordinate
+	}
+
+	keyAlg, signAlg, err := keyTypeToAlgos(a.config.PrivateKeyType, a.config.PrivateKeyBits)
+	if err != nil {
+		return err
+	}
+
+	uid, err := connect.CompactUID()
+	if err != nil {
+		return err
+	}
+	commonName := connect.CACN("aws", uid, a.clusterID, a.isPrimary)
+
+	createInput := acmpca.CreateCertificateAuthorityInput{
+		CertificateAuthorityType: aws.String(pcaType),
+		CertificateAuthorityConfiguration: &acmpca.CertificateAuthorityConfiguration{
+			Subject: &acmpca.ASN1Subject{
+				CommonName: aws.String(commonName),
+			},
+			KeyAlgorithm:     aws.String(keyAlg),
+			SigningAlgorithm: aws.String(signAlg),
+		},
+		RevocationConfiguration: &acmpca.RevocationConfiguration{
+			// TODO support CRL in future when we manage revocation in Connect more
+			// generally.
+			CrlConfiguration: &acmpca.CrlConfiguration{
+				Enabled: aws.Bool(false),
+			},
+		},
+		// uid is unique each PCA we create so use it as an idempotency string. We
+		// don't actually retry on failure yet but might as well!
+		IdempotencyToken: aws.String(uid),
+		Tags: []*acmpca.Tag{
+			{Key: aws.String("consul_cluster_id"), Value: aws.String(a.clusterID)},
+			{Key: aws.String("consul_datacenter"), Value: aws.String(a.datacenter)},
+		},
+	}
+
+	a.logger.Printf("[DEBUG] creating new PCA %s", commonName)
+	createOutput, err := a.client.CreateCertificateAuthority(&createInput)
+	if err != nil {
+		return err
+	}
+
+	// wait for PCA to be created
+	newARN := *createOutput.CertificateAuthorityArn
+	describeInput := acmpca.DescribeCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(newARN),
+	}
+	_, err = a.pollLoop("Private CA", 2*time.Minute, func() (bool, string, error) {
+		describeOutput, err := a.client.DescribeCertificateAuthority(&describeInput)
+		if err != nil {
+			if err.(awserr.Error).Code() != acmpca.ErrCodeRequestInProgressException {
+				a.logger.Printf("[ERR] connect.ca.aws: error describing PCA: %s", err.Error())
+				return true, "", fmt.Errorf("error waiting for PCA to be created: %s", err)
+			}
+		}
+		if *describeOutput.CertificateAuthority.Status == acmpca.CertificateAuthorityStatusPendingCertificate {
+			a.logger.Printf("[DEBUG] connect.ca.aws: new PCA %s is ready"+
+				" to accept a certificate", newARN)
+			a.arn = newARN
+			// We don't need to reload this ARN since we just created it and know what
+			// state it's in
+			a.arnChecked = true
+			return true, "", nil
+		}
+		// Retry
+		return false, "", nil
+	})
+	return err
+}
+
+func (a *AWSProvider) getCACSR() (string, error) {
+	input := &acmpca.GetCertificateAuthorityCsrInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+	}
+	a.logger.Printf("[DEBUG] connect.ca.aws: retrieving CSR for %s", a.arn)
+	output, err := a.client.GetCertificateAuthorityCsr(input)
+	if err != nil {
+		a.logger.Printf("[ERR] connect.ca.aws: error retrieving CSR: %s", err)
```

Comment:
This is another case of logging and returning the same error which may cause duplication.